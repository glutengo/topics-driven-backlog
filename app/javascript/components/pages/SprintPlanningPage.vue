<template>
    <section id="sprint-planning-page">
        <h1>Sprint planning</h1>

        <template v-if="nextSprint">
            <p>
                Planning sprint: {{ nextSprint.name }} <br>
                Start: {{ nextSprint.start_date }} <br>
                End: {{ nextSprint.end_date }}
            </p>

            <table class="table table-striped">
                <thead>
                <tr>
                    <th> </th>
                    <th>Identifier</th>
                    <th>Story</th>
                    <th>Story&nbsp;points</th>
                    <th>Operations</th>
                </tr>
                </thead>
                <tbody v-sortable="{handle: '.js-drag-drop', onEnd: saveOrder}">
                <tr is="story" v-for="story in storiesInSprint" :key="story.id" :data="story" view="planning-sprint" @removeFromSprint="removeFromSprint(story.id)"></tr>
                </tbody>
            </table>
        </template>

        <p v-else>No sprint to plan!</p>

        <h2>Backlog</h2>
        <table class="table table-striped">
            <thead>
                <tr>
                    <th>Identifier</th>
                    <th>Story</th>
                    <th>Story&nbsp;points</th>
                    <th>Operations</th>
                </tr>
            </thead>
            <tbody >
                <tr is="story" v-for="story in backlog" :key="story.id" :data="story" view="planning-backlog" @addToSprint="addToSprint(story.id)"></tr>
            </tbody>
        </table>

        <story-form v-if="showForm" v-model="newStory" @cancel="showForm = false" @submit="save"></story-form>
        <b-button v-else type="button" variant="primary" @click="showForm = true">Add story</b-button>
    </section>
</template>

<script>
    import Story from "../elements/Story.vue";
    import StoryForm from "../forms/StoryForm.vue";
    import Sprint from "../elements/Sprint.vue";
    import * as _ from "lodash";
    import moment from "moment";

    export default {
        components: {Sprint, StoryForm, Story},
        data() {
            return {
                showForm: false,
                newStory: {},
            };
        },
        computed: {
            backlog() {
                return this.$store.getters['stories/all']
                    .filter(s => s.sprint_id === null)
                    .sort((a, b) => a.position - b.position);
            },

            nextSprint() {
                const currentDate = moment().format('YYYY-MM-DD');

                return _.first(this.$store.getters['courses/allSprints']()
                    .filter(s => s.end_date >= currentDate)
                    .sort((a, b) => a.start_date.localeCompare(b.start_date)));
            },

            storiesInSprint() {
                if (this.nextSprint) {
                    return this.$store.getters['stories/find']('sprint_id', this.nextSprint.id)
                            .sort((a, b) => a.position - b.position);
                } else {
                    return [];
                }
            },
        },
        methods: {
            /**
             * Save current form state as new story
             */
            save() {
                this.$store.dispatch('stories/save', {
                    story: this.newStory
                }).then(() => {
                    this.newStory = {};
                });
            },

            /**
             * Save new position of dragged story
             * @param evt
             */
            saveOrder(evt) {
                const story = this.storiesInSprint[evt.oldIndex];

                if (!story) {
                    return;
                }

                this.$store.dispatch('stories/patch', {
                    id: story.id,
                    field: 'position',
                    value: evt.newIndex + 1, // act_as_list is 1-indexed
                    fetch: true,
                });
            },

            /**
             * Add story to next sprint
             * @param storyId
             */
            addToSprint(storyId) {
                if (!this.nextSprint) {
                    return;
                }

                this.$store.dispatch('stories/patch', {
                    id: storyId,
                    field: 'sprint_id',
                    value: this.nextSprint.id,
                });
            },

            /**
             * Remove story from next sprint
             * @param storyId
             */
            removeFromSprint(storyId) {
                this.$store.dispatch('stories/patch', {
                    id: storyId,
                    field: 'sprint_id',
                    value: null,
                });
            }
        }
    };
</script>

<style scoped>

</style>
