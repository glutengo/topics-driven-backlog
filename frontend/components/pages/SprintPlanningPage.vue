<template>
    <section id="sprint-planning-page">
        <h2 class="h4 text-muted" v-if="course">{{ course.title}} </h2>
        <h1>
            Sprint planning
            <small class="text-muted">{{ project.title}}</small>
        </h1>

        <template v-if="nextSprint">
            <p>
                Planning sprint: {{ nextSprint.name }} <br>
                Start: {{ nextSprint.start_date }} <br>
                End: {{ nextSprint.end_date }}
            </p>

            <table class="table table-striped">
                <thead>
                    <tr>
                        <th>&nbsp;</th>
                        <th>Identifier</th>
                        <th>Story</th>
                        <th>Story&nbsp;points</th>
                        <th>Operations</th>
                    </tr>
                </thead>
                <tbody v-sortable="{handle: '.js-drag-drop', onEnd: saveOrder}">
                    <tr 
                        is="StoryItem" 
                        v-for="story in storiesInSprint" 
                        :key="story.id" 
                        :data="story" 
                        view="planning-sprint" 
                        @removeFromSprint="removeFromSprint(story.id)"/>
                </tbody>
            </table>
        </template>

        <p v-else>No sprint to plan!</p>

        <h2>Backlog</h2>
        <table class="table table-striped">
            <thead>
                <tr>
                    <th>Identifier</th>
                    <th>Story</th>
                    <th>Story&nbsp;points</th>
                    <th>Operations</th>
                </tr>
            </thead>
            <tbody >
                <tr 
                    is="StoryItem" 
                    v-for="story in backlog" 
                    :key="story.id" 
                    :data="story" 
                    view="planning-backlog" 
                    @addToSprint="addToSprint(story.id)"/>
            </tbody>
        </table>

        <StoryForm 
            v-if="showForm" 
            v-model="newStory" 
            @cancel="showForm = false" 
            @submit="save"/>

        <BButton 
            v-else 
            type="button" 
            variant="primary" 
            @click="showForm = true">Add story</BButton>
    </section>
</template>

<script>
import * as _ from 'lodash';
import moment from 'moment';
import BButton from '@bootstrap/button/button';
import BaseProjectAwarePage from '@/components/pages/BaseProjectAwarePage';
import StoryItem from '@/components/elements/StoryItem';
import StoryForm from '@/components/forms/StoryForm';
import SprintItem from '@/components/elements/SprintItem';

export default {
    components: { SprintItem, StoryForm, StoryItem, BButton },
    extends: BaseProjectAwarePage,
    data() {
        return {
            showForm: false,
            newStory: this.$store.getters['stories/template'](),
        };
    },
    computed: {
        /**
         * All stories not in sprint, sorted by position
         * @returns {array}
         */
        backlog() {
            return this.project
                ? this.$store.getters['stories/all'](this.project.id)
                    .filter(s => !s.sprint_id)
                    .sort((a, b) => a.position - b.position)
                : [];
        },

        /**
         * Next sprint depending on current date
         * @returns {object}
         */
        nextSprint() {
            const currentDate = moment().format('YYYY-MM-DD');

            return this.course
                ? _.first(
                    this.$store.getters['sprints/all'](this.course.id)
                        .filter(s => s.end_date >= currentDate)
                        .sort((a, b) => a.start_date.localeCompare(b.start_date)))
                : null;
        },

        /**
         * Stories in next sprint
         * @returns {array}
         */
        storiesInSprint() {                        
            return this.nextSprint
                ? this.$store.getters['stories/find'](this.project.id, 'sprint_id', this.nextSprint.id)
                    .sort((a, b) => a.position - b.position)
                : [];
        },
    },
    methods: {
        /**
         * Save current form state as new story
         */
        async save() {
            await this.$store.dispatch('stories/create', {
                parentId: this.project.id,
                ...this.newStory
            });

            this.newStory = this.$store.getters['stories/template']();

            // TODO handle errors in UI
        },

        /**
         * Save new position of dragged story
         * @param {Event} evt
         */
        async saveOrder(evt) {
            const story = this.stories[evt.oldIndex];

            if (!story) {
                return;
            }

            await this.$store.dispatch('stories/update', {
                id: story.id,
                parentId: this.project.id,
                position: evt.newIndex + 1, // act_as_list is 1-indexed
            });

            await this.$store.dispatch('projects/fetch', {
                id: this.project.id,
                parentId: this.course.id
            });

            // TODO handle errors in UI
        },

        /**
         * Add story to next sprint
         * @param {int} storyId
         */
        async addToSprint(storyId) {
            if (!this.nextSprint) {
                return;
            }

            await this.$store.dispatch('stories/update', {
                id: storyId,
                parentId: this.project.id,
                sprint_id: this.nextSprint.id,
            });

            // TODO handle errors in UI
        },

        /**
         * Remove story from next sprint
         * @param {int} storyId
         */
        async removeFromSprint(storyId) {
            await this.$store.dispatch('stories/update', {
                id: storyId,
                parentId: this.project.id,
                sprint_id: null,
            });

            // TODO handle errors in UI

        },
    },
};
</script>
