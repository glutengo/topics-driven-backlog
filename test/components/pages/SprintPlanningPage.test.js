import { mount, createLocalVue } from 'vue-test-utils';
import moment from 'moment';
import Vuex from 'vuex';
import Vuelidate from 'vuelidate';
import SprintPlanningPage from '@/components/pages/SprintPlanningPage';

// take a date far in the future so validations pass
Date.now = jest.fn(() => 7956915742000);

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Vuelidate);

describe('SprintPlanningPage.test.js', () => {
    let cmp, store, actions, storyGetters, courseGetters;

    beforeEach(() => {
        actions = {
            save: jest.fn(),
            patch: jest.fn(),
        };

        storyGetters = {
            all: () => [
                { id: 1, position: 2, status: 0 },
                { id: 2, position: 1, status: 0 },
                { id: 3, position: 3, status: 0, sprint_id: 2 },
                { id: 4, position: 4, status: 1 },
            ],
            find: () => () => [
                { id: 5, position: 5, status: 0, sprint_id: 1 },
                { id: 3, position: 3, status: 0, sprint_id: 1 },
            ],
        };
    });

    describe('with sprints', () => {
        beforeEach(() => {
            courseGetters = {
                allSprints: () => () => [
                    {
                        id: 1,
                        start_date: moment().format('YYYY-MM-DD'),
                        end_date: moment()
                            .add(2, 'days')
                            .format('YYYY-MM-DD'),
                    },
                    {
                        id: 2,
                        start_date: moment()
                            .subtract(1, 'days')
                            .format('YYYY-MM-DD'),
                        end_date: moment()
                            .subtract(1, 'days')
                            .add(2, 'days')
                            .format('YYYY-MM-DD'),
                    },
                ],
            };

            store = new Vuex.Store({
                modules: {
                    stories: {
                        namespaced: true,
                        actions,
                        getters: storyGetters,
                    },
                    courses: {
                        namespaced: true,
                        getters: courseGetters,
                    },
                },
            });

            cmp = mount(SprintPlanningPage, {
                localVue,
                store,
            });
        });

        it('has sorted backlog with stories not in sprints of all statuses', () => {
            expect(cmp.vm.backlog).toEqual([
                { id: 2, position: 1, status: 0 },
                { id: 1, position: 2, status: 0 },
                { id: 4, position: 4, status: 1 },
            ]);
        });

        it('has correct next sprint', () => {
            expect(cmp.vm.nextSprint).toEqual(courseGetters.allSprints()()[1]);
        });

        it('has stories that are in next sprint ordered by position', () => {
            expect(cmp.vm.storiesInSprint).toEqual([
                { id: 3, position: 3, status: 0, sprint_id: 1 },
                { id: 5, position: 5, status: 0, sprint_id: 1 },
            ]);
        });

        it('dispatches save on story save', () => {
            cmp.vm.save();
            expect(actions.save).toBeCalled();
        });

        it('doesnt save the story order if story not found', () => {
            cmp.vm.saveOrder({ oldIndex: 2 });
            expect(actions.patch).not.toBeCalled();
        });

        it('saves the story order on sortable event', () => {
            cmp.vm.saveOrder({ oldIndex: 0, newIndex: 1 });
            expect(actions.patch).toBeCalled();
        });

        it('adds a story to a sprint', () => {
            cmp.vm.addToSprint(1);
            expect(actions.patch).toBeCalled();
        });

        it('adds a story from a sprint', () => {
            cmp.vm.removeFromSprint(1);
            expect(actions.patch).toBeCalled();
        });

        it('has the expected html structure', () => {
            expect(cmp.element).toMatchSnapshot();
        });
    });

    describe('without sprints', () => {
        beforeEach(() => {
            courseGetters = {
                allSprints: () => () => [],
            };

            store = new Vuex.Store({
                modules: {
                    stories: {
                        namespaced: true,
                        actions,
                        getters: storyGetters,
                    },
                    courses: {
                        namespaced: true,
                        getters: courseGetters,
                    },
                },
            });

            cmp = mount(SprintPlanningPage, {
                localVue,
                store,
            });
        });

        it('has no next sprint if no sprints', () => {
            expect(cmp.vm.nextSprint).toBeFalsy();
        });

        it('has no stories in sprint if no sprint', () => {
            expect(cmp.vm.storiesInSprint).toEqual([]);
        });

        it('doesnt a story to a sprint if no sprint', () => {
            cmp.vm.addToSprint(1);
            expect(actions.patch).not.toBeCalled();
        });

        it('has the expected html structure', () => {
            expect(cmp.element).toMatchSnapshot();
        });
    });
});
