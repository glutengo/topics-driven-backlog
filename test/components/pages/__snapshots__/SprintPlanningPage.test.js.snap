// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SprintPlanningPage.test.js with sprints has the expected html structure 1`] = `
<section
  id="sprint-planning-page"
>
  <h1>
    Sprint planning
  </h1>
   
  <p>
    
            Planning sprint:  
    <br />
    
            Start: 2222-02-21 
    <br />
    
            End: 2222-02-23
        
  </p>
   
  <table
    class="table table-striped"
  >
    <thead>
      <tr>
        <th>
           
        </th>
         
        <th>
          Identifier
        </th>
         
        <th>
          Story
        </th>
         
        <th>
          Story points
        </th>
         
        <th>
          Operations
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr
        class="story-item"
      >
        <td>
          <span
            class="js-drag-drop"
          >
            <svg
              aria-label="Drag and drop to change order"
              class="fa-icon"
              height="16"
              role="img"
              version="1.1"
              viewBox="0 0 1792 1792"
              width="16"
            >
              <path
                d="M1792 896q0 26-19 45l-256 256q-19 19-45 19t-45-19-19-45v-128h-384v384h128q26 0 45 19t19 45-19 45l-256 256q-19 19-45 19t-45-19l-256-256q-19-19-19-45t19-45 45-19h128v-384h-384v128q0 26-19 45t-45 19-45-19l-256-256q-19-19-19-45t19-45l256-256q19-19 45-19t45 19 19 45v128h384v-384h-128q-26 0-45-19t-19-45 19-45l256-256q19-19 45-19t45 19l256 256q19 19 19 45t-19 45-45 19h-128v384h384v-128q0-26 19-45t45-19 45 19l256 256q19 19 19 45z"
              />
               
    
              <!---->
            </svg>
          </span>
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-danger btn-sm"
            title="Remove from sprint"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M1408 736v192q0 40-28 68t-68 28h-1216q-40 0-68-28t-28-68v-192q0-40 28-68t68-28h1216q40 0 68 28t28 68z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
      </tr>
      <tr
        class="story-item"
      >
        <td>
          <span
            class="js-drag-drop"
          >
            <svg
              aria-label="Drag and drop to change order"
              class="fa-icon"
              height="16"
              role="img"
              version="1.1"
              viewBox="0 0 1792 1792"
              width="16"
            >
              <path
                d="M1792 896q0 26-19 45l-256 256q-19 19-45 19t-45-19-19-45v-128h-384v384h128q26 0 45 19t19 45-19 45l-256 256q-19 19-45 19t-45-19l-256-256q-19-19-19-45t19-45 45-19h128v-384h-384v128q0 26-19 45t-45 19-45-19l-256-256q-19-19-19-45t19-45l256-256q19-19 45-19t45 19 19 45v128h384v-384h-128q-26 0-45-19t-19-45 19-45l256-256q19-19 45-19t45 19l256 256q19 19 19 45t-19 45-45 19h-128v384h384v-128q0-26 19-45t45-19 45 19l256 256q19 19 19 45z"
              />
               
    
              <!---->
            </svg>
          </span>
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-danger btn-sm"
            title="Remove from sprint"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M1408 736v192q0 40-28 68t-68 28h-1216q-40 0-68-28t-28-68v-192q0-40 28-68t68-28h1216q40 0 68 28t28 68z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
      </tr>
    </tbody>
  </table>
   
  <h2>
    Backlog
  </h2>
   
  <table
    class="table table-striped"
  >
    <thead>
      <tr>
        <th>
          Identifier
        </th>
         
        <th>
          Story
        </th>
         
        <th>
          Story points
        </th>
         
        <th>
          Operations
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr
        class="story-item"
      >
        <!---->
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-primary btn-sm"
            title="Add to sprint"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M1408 736v192q0 40-28 68t-68 28h-416v416q0 40-28 68t-68 28h-192q-40 0-68-28t-28-68v-416h-416q-40 0-68-28t-28-68v-192q0-40 28-68t68-28h416v-416q0-40 28-68t68-28h192q40 0 68 28t28 68v416h416q40 0 68 28t28 68z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
      </tr>
      <tr
        class="story-item"
      >
        <!---->
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-primary btn-sm"
            title="Add to sprint"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M1408 736v192q0 40-28 68t-68 28h-416v416q0 40-28 68t-68 28h-192q-40 0-68-28t-28-68v-416h-416q-40 0-68-28t-28-68v-192q0-40 28-68t68-28h416v-416q0-40 28-68t68-28h192q40 0 68 28t28 68v416h416q40 0 68 28t28 68z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
      </tr>
      <tr
        class="story-item"
      >
        <!---->
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-primary btn-sm"
            title="Add to sprint"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M1408 736v192q0 40-28 68t-68 28h-416v416q0 40-28 68t-68 28h-192q-40 0-68-28t-28-68v-416h-416q-40 0-68-28t-28-68v-192q0-40 28-68t68-28h416v-416q0-40 28-68t68-28h192q40 0 68 28t28 68v416h416q40 0 68 28t28 68z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
      </tr>
    </tbody>
  </table>
   
  <button
    class="btn btn-primary"
    type="button"
  >
    Add story
  </button>
</section>
`;

exports[`SprintPlanningPage.test.js without sprints has the expected html structure 1`] = `
<section
  id="sprint-planning-page"
>
  <h1>
    Sprint planning
  </h1>
   
  <p>
    No sprint to plan!
  </p>
   
  <h2>
    Backlog
  </h2>
   
  <table
    class="table table-striped"
  >
    <thead>
      <tr>
        <th>
          Identifier
        </th>
         
        <th>
          Story
        </th>
         
        <th>
          Story points
        </th>
         
        <th>
          Operations
        </th>
      </tr>
    </thead>
     
    <tbody>
      <tr
        class="story-item"
      >
        <!---->
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-primary btn-sm"
            title="Add to sprint"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M1408 736v192q0 40-28 68t-68 28h-416v416q0 40-28 68t-68 28h-192q-40 0-68-28t-28-68v-416h-416q-40 0-68-28t-28-68v-192q0-40 28-68t68-28h416v-416q0-40 28-68t68-28h192q40 0 68 28t28 68v416h416q40 0 68 28t28 68z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
      </tr>
      <tr
        class="story-item"
      >
        <!---->
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-primary btn-sm"
            title="Add to sprint"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M1408 736v192q0 40-28 68t-68 28h-416v416q0 40-28 68t-68 28h-192q-40 0-68-28t-28-68v-416h-416q-40 0-68-28t-28-68v-192q0-40 28-68t68-28h416v-416q0-40 28-68t68-28h192q40 0 68 28t28 68v416h416q40 0 68 28t28 68z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
      </tr>
      <tr
        class="story-item"
      >
        <!---->
         
        <td>
          
            
        
        </td>
         
        <td>
          
            
        
        </td>
         
        <td>
          
            not estimated
        
        </td>
         
        <!---->
         
        <!---->
         
        <!---->
         
        <td>
          <button
            class="btn btn-primary btn-sm"
            title="Edit"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1536 1792"
              width="13.714285714285714"
            >
              <path
                d="M363 1536l91-91-235-235-91 91v107h128v128h107zM886 608q0-22-22-22-10 0-17 7l-542 542q-7 7-7 17 0 22 22 22 10 0 17-7l542-542q7-7 7-17zM832 416l416 416-832 832h-416v-416zM1515 512q0 53-37 90l-166 166-416-416 166-165q36-38 90-38 53 0 91 38l235 234q37 39 37 91z"
              />
               
    
              <!---->
            </svg>
          </button>
           
          <button
            class="btn btn-primary btn-sm"
            title="Add to sprint"
            type="button"
          >
            <svg
              class="fa-icon"
              height="16"
              role="presentation"
              version="1.1"
              viewBox="0 0 1408 1792"
              width="12.571428571428571"
            >
              <path
                d="M1408 736v192q0 40-28 68t-68 28h-416v416q0 40-28 68t-68 28h-192q-40 0-68-28t-28-68v-416h-416q-40 0-68-28t-28-68v-192q0-40 28-68t68-28h416v-416q0-40 28-68t68-28h192q40 0 68 28t28 68v416h416q40 0 68 28t28 68z"
              />
               
    
              <!---->
            </svg>
          </button>
        </td>
         
        <!---->
      </tr>
    </tbody>
  </table>
   
  <button
    class="btn btn-primary"
    type="button"
  >
    Add story
  </button>
</section>
`;
